feat(TS_CONFIG_PATH): `TS_CONFIG_PATH` environment variable for forcing tsconfig loading

KristianFJones opened the pull request:
This PR adds the option to provide `TS_CONFIG_PATH` environment variable to force the path to the TSConfig file to use for transpiling and paths.

KristianFJones commented:
@yoursunny This is one way to implement this, I could also have it be the path of the actual config/project file.

KristianFJones commented:
I'm working on refactoring my testing infrastructure so I can add a proper test, it's actually how the Type-Checking thing got started. I need a test that ensures that a `tsconfig.json` file is being loaded, but since my testing infrastructure and TS-ESNode itself doesn't actually do typechecking I'm having some tiny issues with that.

yoursunny commented:
> I could also have it be the path of the actual config/project file.

I'd prefer being able to specify full path to `tsconfig.json`, such as `TS_PROJECT=$(pwd)/mk/tsconfig-run.json`.
This allows the flexibility that the file does not have to be named `tsconfig.json`.

The simplest implementation is:
```ts
export function getTSConfig(modulePath: string): CompilerOptions {
  const tsConfigPath = process.env.TS_CONFIG_PATH ?? ts.findConfigFile(modulePath, ts.sys.fileExists);
```

yoursunny commented:
I tested this change along with #80 in [NDNts](https://github.com/yoursunny/NDNts) repository, and they appear to work well without using my custom hacks.

```bash
export TS_CONFIG_PATH=$ROOTDIR/mk/tsconfig-literate.json
node --loader @k-foss/ts-esnode --experimental-specifier-resolution=node literate-temp.ts
```

You may want to rename this PR to reflect the changed environment variable name.

Currently this change does not recognize TSConfig `extends`, so I have to copy compilerOptions into the final TSConfig, which is a minor annoyance.
Maybe `ts.parseJsonConfigFileContent` should be used instead of `ts.readConfigFile`.

KristianFJones commented:
> I tested this change along with #80 in [NDNts](https://github.com/yoursunny/NDNts) repository, and they appear to work well without using my custom hacks.
> 
> ```shell
> export TS_CONFIG_PATH=$ROOTDIR/mk/tsconfig-literate.json
> node --loader @k-foss/ts-esnode --experimental-specifier-resolution=node literate-temp.ts
> ```
> 
> You may want to rename this PR to reflect the changed environment variable name.
> 
> Currently this change does not recognize TSConfig `extends`, so I have to copy compilerOptions into the final TSConfig, which is a minor annoyance.
> Maybe `ts.parseJsonConfigFileContent` should be used instead of `ts.readConfigFile`.

I've renamed the PR and fixed the description. And... I got "extends" working. The current `fs.readFile` is only temp until I get my WIP cache system I'm trying to work towards in place.

KristianFJones commented on src/Utils.ts:
```diff
@@ -17,12 +22,13 @@ export function getTSConfig(modulePath: string): CompilerOptions {
+    const jsonText = await fs.readFile(tsConfigPath);
+    const result = ts.parseJsonText(tsConfigPath, jsonText.toString());
 
-    tsConfigCache = ts.convertCompilerOptionsFromJson(
-      tsConfigFile.compilerOptions,
-      pathDirname(tsConfigPath),
+    tsConfigCache = ts.parseJsonSourceFileConfigFileContent(
+      result,
+      ts.sys,
+      dirname(tsConfigPath),
     ).options;
```
This will likely have a small perf impact until I setup proper caching. Need to finish moving this repo over to using @K-FOSS/TS-ESTests and have proper time differences on each PR/commit. Speed is my main priority with TS-ESNode.

KristianFJones commented:
@yoursunny Release should be live in a few minutes.
